packages/block-editor/src/components/inserter/search-items.js

```diff
@@ -20,7 +24,11 @@ const defaultGetCollection = () => null;
  * @return {Array} Words, extracted from the input string.
  */
 function extractWords( input = '' ) {
-	return noCase( input, {
+	if ( typeof extractedWords[ input ] !== 'undefined' ) {
+		return extractedWords[ input ];
+	}
+
+	extractedWords[ input ] = noCase( input, {
 		splitRegexp: [
 			/([\p{Ll}\p{Lo}\p{N}])([\p{Lu}\p{Lt}])/gu, // One lowercase or digit, followed by one uppercase.
 			/([\p{Lu}\p{Lt}])([\p{Lu}\p{Lt}][\p{Ll}\p{Lo}])/gu, // One uppercase followed by one uppercase and one lowercase.
```

Comment:
This PR is caching rather trivial string operations, stores all the results in memory forever, and it's questionable whether the performance impact is really there. What if we tried, for example, to extract these regexps into constants, so that they are not parsed and constructed on each call? That could also be a significant speedup.

Reply:
While playing with this PR, I noticed that the `extract` and `normalize` functions are called repeatedly not only on the `input` string, but also on properties of the searched `item`s. I'm wondering if the `item` data could be memoized somehow. The `item`s are still the same, so why process them again and again on every search?

I'm not sure though if there's a simple way to achieve this. There's the additional complication that although the `item`s are still the same, the `config` with the extraction functions (`getTitle` etc.) can be variable.

Reply:
TBH, I wasn't sure if the `item`s are perfectly stable - it's fair to say some of them may change over time, modified by filters or lazy-loaded plugins. I'm happy to tinker with this in a follow-up PR.

Reply:
If the items are "inserter items" then they should be fairly stable, block type definitions don't change after the editor is finished booting up.

My idea was that the data that `getItemSearchRank` extracts from an `item` would be cached in a weakmap: `WeakMap<Item, ItemSearchInfo>`. But that depends also on the `config` passed to `searchItems`. That makes it hard.

By the way, I noticed that there is a copy & paste of all this logic in the `edit-site` package, in the `page-patterns` component.

Reply:
Ah, bummer that so much of it is duplicated. Might give us some opportunities for refactoring and reusing though - thanks for spotting it!

Reply:
I've started deduplicating some of the code in https://github.com/WordPress/gutenberg/pull/60218